src/EditorFeatures/CSharpTest/CodeActions/ConvertLocalFunctionToMethod/ConvertLocalFunctionToMethodTests.cs

```diff
@@ -0,0 +1,120 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.CodeRefactorings.ConvertLocalFunctionToMethod;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeActions.ConvertLocalFunctionToMethod
+{
+    public class ConvertLocalFunctionToMethodTests : AbstractCSharpCodeActionTest
+    {
+        protected override CodeRefactoringProvider CreateCodeRefactoringProvider(Workspace workspace, TestParameters parameters)
+            => new CSharpConvertLocalFunctionToMethodCodeRefactoringProvider();
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertLocalFunctionToMethod)]
+        public async Task CaptureRead()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C 
+{
+    void M()
+    {
+        int var = 2;
+        int [||]F()
+        {
+            return var;
+        }
+        F();
+    }
+}",
+@"class C 
+{
+    void M()
+    {
+        int var = 2;
+        F(var);
+    }
+
+    private static int F(int var)
+    {
+        return var;
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertLocalFunctionToMethod)]
+        public async Task CaptureWrite()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C 
+{
+    void M()
+    {
+        int var = 2;
+        void [||]F()
+        {
+            var = 1;
+        }
+        F();
+    }
+}",
+@"class C 
+{
+    void M()
+    {
+        int var = 2;
+        F(ref var);
+    }
+
+    private static void F(ref int var)
+    {
+        var = 1;
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertLocalFunctionToMethod)]
+        public async Task TypeParameters()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C 
+{
+    void M<T1>(int i)
+    {
+        void Local1<T2>()
+        {
+            void Local2<T3>()
+            {
+                void [||]Local3<T4>(T4 arg)
+                {
+                }
+                Local3<byte>(42);
+                Local3(42);
+            }
+        }
+    }
+}",
+@"class C 
+{
+    void M<T1>(int i)
+    {
+        void Local1<T2>()
+        {
+            void Local2<T3>()
+            {
+                Local3<T1, T2, T3, byte>(42);
+                Local3<T1, T2, T3, int>(42);
+            }
+        }
+    }
+
+    private static void Local3<T1, T2, T3, T4>(T4 arg)
```

Comment:
I'd love to see the type parameters reduced to just the ones that are used.

Reply:
Done.